crowbar_framework/app/models/service_object.rb

```diff
@@ -1852,4 +1852,25 @@ def node_changed_roles?(node, old_role, new_role)
     roles_in_new = new_role.elements.keys.select { |r| new_role.elements[r].include?(node) }.sort
     roles_in_old != roles_in_new
   end
+
+  # return true if no other attribute than the ignored has changed. And that the node has no other
+  # role than the roles in only_for_roles
+  def relevant_attributes_changed_if_roles?(node, old_role, new_role, ignore_attributes, only_for_roles)
+    # sort roles of only_for_roles
+    only_for_roles = only_for_roles.sort
+
+    # return true if node has no other roles that ones in the list (only_for_roles)
+    roles_in_old = old_role.elements.keys.select { |r| old_role.elements[r].include?(node) }.sort
+    roles_in_new = new_role.elements.keys.select { |r| new_role.elements[r].include?(node) }.sort
+    return true if only_for_roles != roles_in_old || only_for_roles != roles_in_new
```

Comment:
First, you can ignore `old_role`: we only care about `new_role` here (since we only care about attributes that will impact the next chef run, not a previous run).

Then, I think what we want here is something closer to:

```ruby
require "set"
return true unless Set(roles_in_new).subset?(Set(only_for_roles))
```
That is: we don't need to make sure the two things are equal, but the roles the node have must be a subset of the "allowed" roles (`only_for_roles`).

And btw, if you move to sets, you don't need to sort things anymore ;-)

Reply:
I wasn't sure about old_role. Thank you for the new approach